tiledb/sm/filesystem/vfs.cc

```diff
@@ -811,66 +812,100 @@ Status VFS::terminate() {
 }
 
 Status VFS::ls(const URI& parent, std::vector<URI>* uris) const {
-  if (!init_)
-    return LOG_STATUS(Status_VFSError("Cannot list; VFS not initialized"));
+  auto&& [st, entries] = ls_with_sizes(parent);
+  RETURN_NOT_OK(st);
+
+  for (auto& fs : *entries) {
+    uris->emplace_back(fs.path().native());
+  }
+
+  return Status::Ok();
+}
+
+tuple<Status, optional<std::vector<directory_entry>>> VFS::ls_with_sizes(
+    const URI& parent) const {
+  if (!init_) {
+    auto st = LOG_STATUS(Status_VFSError("Cannot list; VFS not initialized"));
+    return {st, std::nullopt};
+  }
 
   // Noop if `parent` is not a directory, do not error out.
   // For S3, GCS and Azure, `ls` on a non-directory will just
   // return an empty `uris` vector.
   if (!(parent.is_s3() || parent.is_gcs() || parent.is_azure())) {
     bool flag = false;
-    RETURN_NOT_OK(is_dir(parent, &flag));
-    if (!flag)
-      return Status::Ok();
+    RETURN_NOT_OK_TUPLE(is_dir(parent, &flag), nullopt);
+
+    if (!flag) {
+      return {Status::Ok(), std::vector<directory_entry>()};
+    }
   }
 
-  std::vector<std::string> paths;
+  optional<std::vector<directory_entry>> entries;
   if (parent.is_file()) {
 #ifdef _WIN32
-    RETURN_NOT_OK(win_.ls(parent.to_path(), &paths));
-#else
-    RETURN_NOT_OK(posix_.ls(parent.to_path(), &paths));
-#endif
-  } else if (parent.is_hdfs()) {
-#ifdef HAVE_HDFS
-    RETURN_NOT_OK(hdfs_->ls(parent, &paths));
+    Status st;
+    std::tie(st, entries) = win_.ls_with_sizes(parent);
```

Comment:
I know it's not your fault, but this kind of code is why I hate the preprocessor. This code would all be much clearer were it to use `if constexpr`. It's out of scope for this PR, in all likelihood.

Reply:
One immediate improvement would be to have an abstract class for the backends we support. This is something I thought about when I observed that `::ls` based on `::ls_with_sizes` looks identical in all backends, thus duplicated code.
An abstract for this should have a virtual name() to get the log messages right and of course we need to unify the APIs backends expose (right now there are slight differences). This if I got it correctly should help us at least move all the `#ifdef PLATFORM`s in a single place. `VFS` would call `backend_->api_call(args)` and `LOG_STATUS` using `backend_->name()`. This is a tedious task though, I'd love not to have to do it in this PR.

Reply:
Absolutely no worries, it's good we see a potential improvement to make.